### ISSUE TITLE
On firefox restart, tabs are a mix from most recently viewed group and some other group, then switches to the other group

### ISSUE THREAD

**Issue by slinkp** [Quicksaver/Tab-Groups#411]
Running Firefox 48.0 for Ubuntu.

1. Take note of which group we're viewing, let's call it "Group A"
2. Quit Firefox
3. restart computer (not sure if necessary)
4. Start Firefox

Expected:
Firefox opens with the last used tab group A.

Actual behavior:

1. The correct tab group name (call it "Group A") is shown at top right
2. All but one of the tab names are for tabs in that correct group A
3. One of the tabs - the currently focused tab, let's call it "Tab X"  - is not from group A; it is from a totally different "group B".   Tab X and group B may be ones I have not viewed for days or even weeks.
4. If I wait for Tab X to fully load, and then click on the tab groups menu at top right (but don't select a group), then the tab group name at top right switches to "group B".  But the tabs shown are still the "group A" tabs plus the errant "tab X".
6. If I go to the tab group dropdown and switch back to group A, everything works as expected: only tabs from group A are loaded.  If I select group B, then only tabs from group B are loaded.

Not sure, but this may be related: Several times when this happens I get a "Script is not responding" dialog from Firefox. Not clear which page it corresponds to, but it may be the currently viewed tab X that opened from the wrong group B?

**Comment by slinkp** [comment]
Updated description with corrections

**Comment by Jakob-H** [comment]
I often see a similar behavior. Also in combination with the "Script is not responding" dialog.

Firefox for windows. I'm not sure if it still occurs with Firefox 49, but I think it still does.
I've got between 800 and 1000 Tabs open and only 4GB ram. I think this makes starting Firefox very slow and often strange things happen regarding tab-groups. I can't remember a day without a "Script is not responding"-dialog.

I have many group, but most of the time I don't switch groups for hours. In most sessions I don't use more than 10% of my groups. And I hardly ever used more than 10% of my open tabs during one session. **So I would appreciate very much, if less tabs (or less tab-content) would be loaded on Firefox startup** (especially on systems with low ram). Of course the names of open tabs in the current group must be loaded from the beginning, so that I can see which tabs are open. But I would like that tabs from other groups are not loaded before I choose other groups.

**Comment by Quicksaver** [comment]
> I would like that tabs from other groups are not loaded before I choose other groups.

This already happens, by default tabs are only loaded once you have clicked on them at least once to select them. Can you go to about:config and confirm that ```browser.sessionstore.restore_on_demand``` is * true*?

**Comment by Jakob-H** [comment]
> Can you go to about:config and confirm that browser.sessionstore.restore_on_demand is * true*?
Yes, it is "true" (standard)

Is it possible that the session manager contributes to the problem?

**Comment by Quicksaver** [comment]
Possibly, see #286.

**Comment by Jakob-H** [comment]
> Possibly, see #286.

Yes, I think #286 describes many problem that I have experienced too.

@slinkp Do you have session manager installed too?
> restart computer (not sure if necessary)

My feeling is that you have to wait something like 10 minutes after closing Firefox before you restart your computer or before you restart Firefox, so that Firefox gets enough time to save the current session properly.
I've got this feeling, because after I close Firefox the percentage of used ram stays at ~90% for some minutes after I close Firefox. And my feeling is that I have especially many problems on the next Firefox start, when I didn't wait long enough after the last closing of Firefox.

**Comment by Quicksaver** [comment]
Tab Groups plays no part in saving the session on shutdown, that's all Firefox. The groups information is always kept updated along the session, so the add-on doesn't hold shutdown at all by itself. And this groups info is really barely more than a numerical identifier to each tab to know to which group it's saved; this should be irrelevant in session size or saving time. If you think that it's taking too long to finish shutting down, it might be worth contacting Firefox's devs directly about it.

The add-on also shouldn't hold up startup at all. The (in)visible tabs are properly restored by Firefox itself, because that info is already in the session that's loaded. Tab Groups itself doesn't initialize most of itself until the first time you click on the button or enter groups view.

In either case, you should start by trying to disable all your other add-ons except for Tab Groups, to see if any of them, whether by themselves or through an incompatibility with Tab Groups, is causing any of these issues.

I would also suggest creating a backup of your tabs and groups through the add-on's preferences, then [create an entirely new profile](https://support.mozilla.org/kb/profile-manager-create-and-remove-firefox-profiles) and install Tab Groups there and import back your tabs and group from your backup into it, to see if you get similar results.

Another entirely new test is simply disable Tab Groups, but before you restart Firefox make sure either Firefox or Session Manager is set to restore your last session. Then see if you still have huge startup and shutdown times like these. Note that by disabling the add-on you don't actually lose your groups, they're all there and all their tabs (as long as you keep your session of course), you just won't have any way to access them until you re-enable the add-on.

**Comment by Quicksaver** [comment]
@Jakob-H also, if you do have anything more to add, please open a new issue and post there. We've detoured from the original issue here, which is completely different, and I didn't even notice.

@slinkp I'm sorry for diverting there. Have you tried closing all your other add-ons except for Tab Groups and see if that still happens then?

**Comment by Quicksaver** [comment]
@slinkp [beta version 2.1b2](https://github.com/Quicksaver/Tab-Groups/releases/tag/v2.1b2) fixes a similar issue, although I'm not entirely sure it's the same you are having. If you could please confirm if you still have this issue with that beta, I would appreciate it.

**Comment by slinkp** [comment]
@Quicksaver thank you. I am now running the latest beta 2.1b3.
Here's what I've noticed:
* With my default add-ons enabled, I still see *some* of the behavior mentioned in this bug:
 * I get the "unresponsive script" warning at startup
 * this warning may come from a page that wasn't on the most recently viewed group
 * in that case, i seem to get switched to the group containing the unresponsive page
 * I have NOT seen the mix of tabs from different groups (yet).  So far, each time I start, I either stay on the expected (last viewed) group, or get switched entirely to a different group.
* If I disable all other add-ons, I do not see any of these problems at all.
 * Also, startup and shutdown are MUCH faster, and Firefox subjectively feels faster overall (can't prove it but I feel like I'm now spending little or no time waiting for unresponsive UI).

I'll see if I can deduce which combination of add-ons triggers the problem.

I have recently noticed that the unresponsive script usually (always?) seems to be something from Lastpass, so I'll especially watch for whether Lastpass is involved in this problem.

**Comment by slinkp** [comment]
Also, somebody above asked whether I have Session Manager installed. I do not.

**Comment by Quicksaver** [comment]
> - I get the "unresponsive script" warning at startup
- this warning may come from a page that wasn't on the most recently viewed group
- in that case, i seem to get switched to the group containing the unresponsive page

If this unresponsive script warning switches your selected tab at startup before Tab Groups has finished its early initialization, then it's entirely possible for this tab to be shown mixed in with tabs from the current group; Tab Groups will just assume that tab is being shown as expected, it doesn't even check to see which group it is from because it just assumes that whatever tabs you have visible "currently" are correct because those were (or, of course, normally should have been) visible when you last shut down Firefox.

Knowing this, do you think this is what could be happening?

Note that a correct check for all tabs would require Tab Groups to fully initialize at startup instead of only at the first time you click on its button to switch groups, and I would avoid doing this because it would cause your startup to become even slower (a lot). So if that really is the cause, I'm not sure there's much I can do.

> - If I disable all other add-ons, I do not see any of these problems at all.
  - Also, startup and shutdown are MUCH faster, and Firefox subjectively feels faster overall (can't prove it but I feel like I'm now spending little or no time waiting for unresponsive UI).

I suggest you try to find out if there's a specific one that's causing most of the slower startups and the unresponsive script warnings. LastPass is a possibility yeah, in my experience it always seems to be a fairly heavier add-on than it's worth or even expected to be, but that may be just my biased opinion, so I'd follow up with their support to be sure; I'm not entirely sure they have many users with 800 tabs like that, so I'm sure they'll appreciate the input so they can make performance improvements for those cases like yours if necessary. ;)

Please do let me know if you think there's a specific add-on that's not playing well with Tab Groups and could be causing the issues though.

Also, please also try [beta version 2.1b4](https://github.com/Quicksaver/Tab-Groups/releases/tag/v2.1b4), I made an oversight in the last beta I linked you to, and this new beta should be fairly more responsive on startup (although it shouldn't have much of a difference from the last 2.0 release or in your case at all even, but who knows).

**Comment by slinkp** [comment]
I am a Lastpass premium subscriber; filed a ticket with them, will see if I get anywhere.

**Comment by slinkp** [comment]
So, yes, disabling Lastpass makes the symptoms go away.

Following a tip I got from Lastpass support, I am trying out multiprocess on firefox, aka Electrolysis.
I followed these instructions https://wiki.mozilla.org/Electrolysis#Force_Enable to get it working.
That done, I no longer notice performance problems with lastpass + tab groups combined in normal usage - I hadn't reallized how slow the combination was making Firefox in general until I tried disabling them per your suggestions! . So I'm glad to know about that workaround, but I'm not sure this has resolved the startup problems I reported here though; I still see the "unresponsive script" warning, and I did notice the unexpected switching-to-other-group behavior at least once today.

I mentioned all this in an update on the issue that I filed w/ Lastpass too, in case it helps them.

I've just installed beta 2.1b8 of Tab Groups.  Will report back if things seem any better, worse, or same.

**Comment by Quicksaver** [comment]
Before anything, I'm sorry I can't give you a more concrete fix for now, I realize it can be annoying having it switch like this. :(

> I still see the "unresponsive script" warning, and I did notice the unexpected switching-to-other-group behavior at least once today.

To clarify, tabs don't mix between groups, it just jumps from one group to another, is this correct?

If yes, then that's "expected". If it's switching to a hidden tab (from another group), then it's following through with switching the group as well. Obviously the add-on can't "prevent" tab switching at all, because it wouldn't know which tab switches were "valid"; technically there's just a "tab switched" notification, there's no usable information about why or how this happened. So correctly switching groups when necessary is pretty much the only thing the add-on can do when this happens.

What's weird is that you shouldn't get that unresponsive script warning from tabs that haven't been loaded yet on startup. Have you also checked if ```browser.sessionstore.restore_on_demand``` is ```true``` in about:config? If it is true, something is trying to load those tabs (possibly LastPass?), and if every tab you have is being loaded at startup then it's really no surprise it's causing all kinds of slowness and unresponsive scripts warnings.

**Comment by Quicksaver** [comment]
Alright, I finally gave LastPass a try and opened a lot of tabs. I got it to reproduce everything exactly as you mentioned.

On the one hand, LastPass doesn't seem to be loading all the tabs at startup like I thought it was.

But it does seem to do something with all those tabs. This happens regardless of Tab Groups though, I disabled Tab Groups, made sure I kept all my tabs by going to Firefox's options and setting "When Firefox starts" to "Show my windows and tabs from last time" (even tabs from other groups are kept this way, they just aren't easily accessible because they're hidden) and after restarting Firefox I still get the same problem, except now groups don't switch of course because Tab Groups isn't there to do it, it just mixes the tab it switches to with whatever tabs were already visible; so your initial problem I believe.

Actually, I can get this problem by restarting Firefox with LastPass disabled, then enable it only after Firefox is already fully started. I get unresponsive script warnings and random tab switching like before as well. The time LastPass takes to initialize is perfectly distinctive there as well.

All in all, this is a problem with LastPass, it definitely needs lots of performance improvements when initializing a large number of tabs, even unloaded ones. Even not accounting for this unresponsive script problem, my startup time with LastPass increased over five-fold. In my opinion, that's just wrong, especially because it's trying to do something with tabs that don't really exist / aren't loaded yet. I wonder what it's trying to do there...

**Comment by slinkp** [comment]
Thanks, that's interesting. Yes, I did have `browser.sessionstore.restore_on_demand` set `true`.

Re. this: 

> To clarify, tabs don't mix between groups, it just jumps from one group to another, is this correct?

Yes, correct. When I first reported this issue, I was seeing what looked like a tab mixed into the wrong group, as discussed earlier. That part of the symptom seems to be gone now.  Now I just pretty quickly end up in an unexpected group.

I also confirm the same behavior (unresponsive script warning + switching to a random tab) if I start with LastPass disabled and then enable it.

**Comment by Quicksaver** [comment]
Then you should definitely bring this up with LastPass support, if there's a reason that LastPass needs to use so many resources on unloaded tabs they would know more easily than me. And of course if there's no reason for that to happen, then its startup process definitely needs some improvement with a large number of tabs. ;)

**Comment by slinkp** [comment]
Yup I already have a ticket open over there. Thanks!

**Comment by Quicksaver** [comment]
If you can, please let me know if anything comes of it, I'm interested in what they do about this.
